ktor-network/posix/src/io/ktor/network/sockets/DatagramSocketImpl.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.network.util.*
+import io.ktor.util.network.*
+import io.ktor.utils.io.*
+import io.ktor.utils.io.core.*
+import io.ktor.utils.io.errors.*
+import kotlinx.cinterop.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.channels.*
+import platform.posix.*
+import kotlin.coroutines.*
+
+internal class DatagramSocketImpl(
+    descriptor: Int,
+    selector: SelectorManager,
+    private val _localAddress: NetworkAddress,
+    private val _remoteAddress: NetworkAddress?,
+    parent: CoroutineContext = EmptyCoroutineContext
+) : BoundDatagramSocket, ConnectedDatagramSocket, ReadWriteSocket,
+    BaseNativeSocket(descriptor, selector, parent) {
+
+    override val localAddress: NetworkAddress
+        get() = _localAddress
+    override val remoteAddress: NetworkAddress
+        get() = _remoteAddress!! // TODO: What should happen here?
+
+    private val sender = BroadcastChannel<Datagram>(Channel.BUFFERED)
+
+    private val receiver = produce<Datagram>(coroutineContext) {
+        while (true) {
+            val received = receiveImpl()
+            channel.send(received)
+        }
+    }
+
+    init {
+        makeShared()
+
+        launch {
+            sender.consumeEach { datagram ->
+                sendImpl(datagram)
+            }
+        }
+    }
+
+    private tailrec suspend fun sendImpl(
+        datagram: Datagram,
+        bytes: ByteArray = datagram.packet.readBytes()
+    ) {
+        var result: Int? = null
+        bytes.usePinned { pinned ->
+            datagram.address.address.nativeAddress { address, addressSize ->
+                result = sendto(
+                    descriptor,
+                    pinned.addressOf(0),
+                    bytes.size.convert(),
+                    0,
+                    address,
+                    addressSize
+                ).toInt()
+            }
+        }
+        result!!
+
+        when (result!!) {
+            0 -> {
```

Comment:
Could you extract constants with names here?

Reply:
It looks like a duplicate and we can extract this in a separate method. Probably we already have one more specific, but we can try reusing it

Reply:
I am not sure what exactly you would like me to extract in a separate method here. There are different actions done based on the result in every place where result is checked. The errors are already handled by `PosixException.forErrno()`.

Reply:
Made some changes to the naming of `result`, does this make more sense? I can revert if needed. https://github.com/ktorio/ktor/pull/2129/commits/56d25383602d33cc1b792776fe4374f7572418f9

Reply:
> Could you extract constants with names here?

Yes, sure, I could extract the `0` and `-1`. However, I looked into the Ktor code but could not find any other places where these constants were extracted, so I am not sure what to name them or if it would make sense. The `-1` could maybe be named `POSIX_ERROR` or so but I am not sure what to name the `0`.